x/valset-pref/simulation/sim_msgs.go

```diff
@@ -31,88 +31,123 @@ func RandomMsgSetValSetPreference(k valsetkeeper.Keeper, sim *osmosimtypes.SimCt
-	// check if the delegator valset created
-	err := GetRandomExistingValSet(ctx, k, sim, delegator.Address)
+	delAddr := delegator.Address
+
+	// get delegator valset preferences
+	preferences, err := k.GetDelegationPreferences(ctx, delAddr.String())
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("no delegations found")
 	}
 
-	// check that the delegator has delegated tokens
-	_, err = GetRandomExistingDelegation(ctx, k, sim, delegator.Address)
+	delegation := preferences.Preferences[rand.Intn(len(preferences.Preferences))]
+	val, err := sdk.ValAddressFromBech32(delegation.ValOperAddress)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("validator address not formatted")
+	}
+
+	validator, found := sim.StakingKeeper().GetValidator(ctx, val)
+	if !found {
+		return nil, fmt.Errorf("Validator not found")
+	}
+
+	// check if the user has delegated tokens to the valset
+	del, found := sim.StakingKeeper().GetDelegation(ctx, delAddr, val)
+	if !found {
+		return nil, fmt.Errorf("user hasn't delegated tokens to the validator, %s", val.String())
+	}
+
+	totalBond := validator.TokensFromShares(del.GetShares()).TruncateInt()
+	if !totalBond.IsPositive() {
+		return nil, fmt.Errorf("%s is not present", sdk.DefaultBondDenom)
 	}
 
-	undelegationCoin := sim.RandExponentialCoin(ctx, delegator.Address)
+	undelegationCoin := rand.Intn(int(totalBond.Int64()))
+
 	return &types.MsgUndelegateFromValidatorSet{
-		Delegator: delegator.Address.String(),
-		Coin:      undelegationCoin,
+		Delegator: delAddr.String(),
+		Coin:      sdk.NewCoin(sdk.DefaultBondDenom, sdk.NewInt(int64(undelegationCoin))),
 	}, nil
 }
 
 func RandomMsgReDelegateToValSet(k valsetkeeper.Keeper, sim *osmosimtypes.SimCtx, ctx sdk.Context) (*types.MsgRedelegateValidatorSet, error) {
+	// random delegator account
 	delegator := sim.RandomSimAccount()
-	// check if the delegator valset created
-	err := GetRandomExistingValSet(ctx, k, sim, delegator.Address)
-	if err != nil {
-		return nil, err
-	}
+	delAddr := delegator.Address
 
-	// check that the delegator has delegated tokens
-	_, err = GetRandomExistingDelegation(ctx, k, sim, delegator.Address)
+	// existing delegations
+	delegations, err := k.GetDelegationPreferences(ctx, delAddr.String())
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("no delegations found")
 	}
 
-	remainingWeight := sdk.NewDec(1)
-	preferences, err := GetRandomValAndWeights(ctx, k, sim, remainingWeight)
-	if err != nil {
-		return nil, err
-	}
+	for _, dels := range delegations.Preferences {
+		val, err := sdk.ValAddressFromBech32(dels.ValOperAddress)
+		if err != nil {
+			return nil, fmt.Errorf("validator address not formatted")
+		}
 
-	return &types.MsgRedelegateValidatorSet{
-		Delegator:   delegator.Address.String(),
-		Preferences: preferences,
-	}, nil
-}
+		if sim.StakingKeeper().HasReceivingRedelegation(ctx, delAddr, val) {
+			return nil, fmt.Errorf("receveing redelegation is not allowed for source validators")
```

Comment:
what does it mean here by "source validators"?

Reply:
For redelegation, we need validators to redelegate to(target validators) and validators to redelegate from(source validators).